system/HTTP/Parameters/ParametersInterface.php

```diff
@@ -0,0 +1,67 @@
+<?php
+
+declare(strict_types=1);
+
+/**
+ * This file is part of CodeIgniter 4 framework.
+ *
+ * (c) CodeIgniter Foundation <admin@codeigniter.com>
+ *
+ * For the full copyright and license information, please view
+ * the LICENSE file that was distributed with this source code.
+ */
+
+namespace CodeIgniter\HTTP\Parameters;
+
+use Countable;
+use IteratorAggregate;
+
+/**
+ * @template TKey of string
+ * @template TValue
+ *
+ * @extends IteratorAggregate<TKey, TValue>
+ */
+interface ParametersInterface extends IteratorAggregate, Countable
+{
+    /**
+     * @param array<TKey, TValue> $parameters
+     */
+    public function __construct(array $parameters = []);
+
+    /**
+     * @param array<TKey, TValue> $parameters
+     */
+    public function override(array $parameters = []): void;
+
+    /**
+     * @param TKey $key
+     */
+    public function has(string $key): bool;
+
+    /**
+     * @param TKey   $key
+     * @param TValue $default
+     *
+     * @return TValue|null
+     */
```

Comment:
I understand now the issue. You need to specify another template for this default. Like:

```suggestion
     * @template TDefault
     *
     * @param TKey   $key
     * @param TDefault $default
     *
     * @return TValue|TDefault|null
     */
```

Reply:
That's the problem.
I didn't plan on `null` in scalar data. `$_GET['a'] = null` from the request will never exist. Yes, this is acceptable in other places.

Are there any examples for `null` tests?

Reply:
This is the tinkering I made to allow strict default value and not allowing `null` for InputParameters.

```diff
diff --git a/system/HTTP/Parameters/InputParameters.php b/system/HTTP/Parameters/InputParameters.php
index 8aa0508075..a45f7b22bd 100644
--- a/system/HTTP/Parameters/InputParameters.php
+++ b/system/HTTP/Parameters/InputParameters.php
@@ -19,8 +19,9 @@ use CodeIgniter\Exceptions\RuntimeException;
 /**
  * @template TKey of string
  * @template TValue of scalar|array<(int|string), mixed>
+ * @template TDefault of scalar|array<(int|string), mixed>
  *
- * @extends Parameters<TKey, TValue>
+ * @extends Parameters<TKey, TValue, TDefault>
  *
  * @see \CodeIgniter\HTTP\Parameters\InputParametersTest
  */
@@ -35,7 +36,7 @@ class InputParameters extends Parameters
         }
     }
 
-    public function get(string $key, mixed $default = null): mixed
+    public function get(string $key, mixed $default = ''): mixed
     {
         if ($default !== null && ! is_scalar($default)) {
             throw new InvalidArgumentException(sprintf('The default value for the InputParameters must be a scalar type, "%s" given.', gettype($defau
lt)));
diff --git a/system/HTTP/Parameters/Parameters.php b/system/HTTP/Parameters/Parameters.php
index 20f9b7d66c..8629f750d2 100644
--- a/system/HTTP/Parameters/Parameters.php
+++ b/system/HTTP/Parameters/Parameters.php
@@ -19,8 +19,9 @@ use CodeIgniter\Exceptions\RuntimeException;
 /**
  * @template TKey of string
  * @template TValue
+ * @template TDefault
  *
- * @implements ParametersInterface<TKey, TValue>
+ * @implements ParametersInterface<TKey, TValue, TDefault>
  *
  * @see \CodeIgniter\HTTP\Parameters\ParametersTest
  */
diff --git a/system/HTTP/Parameters/ParametersInterface.php b/system/HTTP/Parameters/ParametersInterface.php
index 6873e78f78..21ccd09744 100644
--- a/system/HTTP/Parameters/ParametersInterface.php
+++ b/system/HTTP/Parameters/ParametersInterface.php
@@ -19,6 +19,7 @@ use IteratorAggregate;
 /**
  * @template TKey of string
  * @template TValue
+ * @template TDefault
  *
  * @extends IteratorAggregate<TKey, TValue>
  */
@@ -41,9 +42,9 @@ interface ParametersInterface extends IteratorAggregate, Countable
 
     /**
      * @param TKey   $key
-     * @param TValue $default
+     * @param TDefault $default
      *
-     * @return TValue|null
+     * @return TValue|TDefault|null
      */
     public function get(string $key, mixed $default = null): mixed;
 
diff --git a/system/HTTP/Parameters/ServerParameters.php b/system/HTTP/Parameters/ServerParameters.php
index c6e87d49ae..0bbf8f2b07 100644
--- a/system/HTTP/Parameters/ServerParameters.php
+++ b/system/HTTP/Parameters/ServerParameters.php
@@ -16,8 +16,9 @@ namespace CodeIgniter\HTTP\Parameters;
 /**
  * @template TKey of string
  * @template TValue
+ * @template TDefault
  *
- * @extends Parameters<TKey, TValue>
+ * @extends Parameters<TKey, TValue, TDefault>
  */
 class ServerParameters extends Parameters
 {
```